File: FinalReview/lpe_python/eval.py
'''
CFG
E -> M + E|M - E|M
M -> N * M|N / M|sq M|N
N -> n|(E)
where n is any integer
'''
from lpe_python.lexer import lex
from lpe_python.parser import parser

class Node:
  def __init__(self,v,t=None,left=None,right=None):
    self.type = t
    self.value = v
    self.left = left
    self.right = right

def eval(tree):
  if tree.type:
    leftv = eval(tree.left)
    if tree.value == "sq":
      return leftv * leftv
    else:
      rightv = eval(tree.right)
    if tree.value == "+":
      rightv = eval(tree.right)
      return leftv + rightv
    if tree.value == "-":
      return leftv - rightv
    if tree.value == "*":
      return leftv * rightv
    if tree.value == "/":
      return leftv / rightv
  return int(tree.value)

File: FinalReview/lpe_ocaml/eval.ml
type node = { value: string; typ: string option; left: node option; right: node option }

let rec eval_node tree =
  match tree.typ with
  | Some t ->
    let leftv = eval_node tree.left in
    (match tree.value with
     | "sq" -> leftv * leftv
     | "+" ->
       let rightv = eval_node tree.right in
       leftv + rightv
     | "-" ->
       let rightv = eval_node tree.right in
       leftv - rightv
     | "*" ->
       let rightv = eval_node tree.right in
       leftv * rightv
     | "/" ->
       let rightv = eval_node tree.right in
       leftv / rightv
     | _ -> failwith "Invalid operation"
    )
  | None -> int_of_string tree.value

let eval tree = eval_node tree
